### PR TITLE
add `(none)` in -getinfo `Warnings:` if no warning returned

### PR DESCRIPTION
**Pull request by ghost**
Adds `(none)` in warnings when no warnings returned by -getinfo

Reviewers can test this by making the following change in `/src/warnings.cpp`:

```diff
bilingual_str GetWarnings(bool verbose)
{
    bilingual_str warnings_concise;
    std::vector<bilingual_str> warnings_verbose;

    LOCK(g_warnings_mutex);

    // Pre-release build warning
    if (!CLIENT_VERSION_IS_RELEASE) {
-        warnings_concise = _("This is a pre-release test build - use at your own risk - do not use for mining or merchant applications");;
+        warnings_concise = _("");;
```

Before this pull request:

```
$ bitcoin-cli -getinfo
Chain: regtest
Blocks: 0
Headers: 0
Verification progress: 100.0000%
Difficulty: 4.656542373906925e-10

Network: in 0, out 0, total 0
Version: 239900
Time offset (s): 0
Proxies: n/a
Min tx relay fee rate (BTC/kvB): 0.00001000

Warnings:
```

After this pull request:

```diff
$ bitcoin-cli -getinfo
Chain: regtest
Blocks: 0
Headers: 0
Verification progress: 100.0000%
Difficulty: 4.656542373906925e-10

Network: in 0, out 0, total 0
Version: 239900
Time offset (s): 0
Proxies: n/a
Min tx relay fee rate (BTC/kvB): 0.00001000

Warnings: (none)
```

### CODE REVIEW

**Review by jessebarton — APPROVED** [review]
Concept ACK [1249aa6](https://github.com/bitcoin/bitcoin/pull/24632/commits/1249aa64d56a038381c59deb19dc9daec9422f69)

Confirmed Warning is being thrown even if there are no Warnings
The version is showing 220000 I think because of the version of bitcoind im running, I could be wrong though.

```Chain: main
Blocks: 728446
Headers: 728446
Verification progress: 99.9998%
Difficulty: 27452707696466.39

Network: in 8, out 10, total 18
Version: 220000
Time offset (s): 0
Proxies: n/a
Min tx relay fee rate (BTC/kvB): 0.00001000

Warnings:
```
Compiled and tested fix with no issues with the desired outcome of not throwing empty warnings.

```
Chain: main
Blocks: 728448
Headers: 728448
Verification progress: 99.9999%
Difficulty: 27452707696466.39

Network: in 15, out 10, total 25
Version: 220000
Time offset (s): 0
Proxies: n/a
Min tx relay fee rate (BTC/kvB): 0.00001000
```

**Comment by 1440000bytes** [comment]
> The version is showing 220000 I think because of the version of bitcoind im running, I could be wrong though.

@jessebarton I see `Version: 230000` when running RC2. It prints server version from `getnetworkinfo` so you are right that its 220000 because of bitcoind in your output.

File: src/bitcoin-cli.cpp [hunk 1]
```diff
@@ -1059,7 +1059,10 @@ static void ParseGetInfoResult(UniValue& result)
         result_string += "\n";
     }
 
-    result_string += strprintf("%sWarnings:%s %s", YELLOW, RESET, result["warnings"].getValStr());
+    if (result["warnings"].getValStr()!="")
+    {
+        result_string += strprintf("%sWarnings:%s %s", YELLOW, RESET, result["warnings"].getValStr());
+    }
```
Comment (jonatack) [inline src/bitcoin-cli.cpp:1065]: suggestions

- don't add extra line breaks when no warning is present
- avoid running getValStr() twice
- use empty()
- clang-format of brackets in the conditional

```diff
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -1038,7 +1038,7 @@ static void ParseGetInfoResult(UniValue& result)
         result_string += strprintf("Transaction fee rate (-paytxfee) (%s/kvB): %s\n\n", CURRENCY_UNIT, result["paytxfee"].getValStr());
     }
     if (!result["balance"].isNull()) {
-        result_string += strprintf("%sBalance:%s %s\n\n", CYAN, RESET, result["balance"].getValStr());
+        result_string += strprintf("%sBalance:%s %s\n", CYAN, RESET, result["balance"].getValStr());
     }
 
     if (!result["balances"].isNull()) {
@@ -1056,13 +1056,13 @@ static void ParseGetInfoResult(UniValue& result)
                                        result["balances"][wallet].getValStr(),
                                        wallet.empty() ? "\"\"" : wallet);
         }
-        result_string += "\n";
     }
 
-    if (result["warnings"].getValStr()!="")
-    {
-        result_string += strprintf("%sWarnings:%s %s", YELLOW, RESET, result["warnings"].getValStr());
+    const std::string warnings{result["warnings"].getValStr()};
+    if (!warnings.empty()) {
+        result_string += strprintf("\n%sWarnings:%s %s", YELLOW, RESET, warnings);
     }
+
     result.setStr(result_string);
 }
```
With this suggestion, I think there is still an extra line break after the balance/balances, if you can look into it.

**Comment by jonatack** [comment]
Concept ACK. Previously, -getinfo returned a JSON-RPC-like response with the "warnings" field returned by getnetworkinfo after calling `GetWarnings()` in `src/warnings.{h,cpp}`. With #21832, -getinfo was converted to a human-readable format without taking into account that the field, when empty, could appear odd. This was likely overlooked because in pre-release development the tester will always see `Warnings: This is a pre-release test build - use at your own risk - do not use for mining or merchant applications`.

File: src/bitcoin-cli.cpp (hunk 1 above)
Comment (1440000bytes) [reply src/bitcoin-cli.cpp:1064]: @jonatack thanks for review and suggestion. Made changes, tested and it looks good. Let me know if there are any issues with line breaks:

With warning: https://pastebin.com/raw/eN3SGjiX
Without warning: https://pastebin.com/raw/frfnqfvU

Comment (jonatack) [reply src/bitcoin-cli.cpp:1064]: As I mentioned, there is still an extra line, and the commits should be squashed, as the extra line is created by the changes here.

Comment (1440000bytes) [reply src/bitcoin-cli.cpp:1064]: Done. Removed extra lines: https://pastebin.com/raw/bZk4YNj5

**Comment by jonatack** [comment]
Thanks for updating.

- It looks like there is now (a) an extra line between "Balances" and the balances, and (b) a line missing before the warnings, with both one balance and multiple balances. Example:

```
₿ ./src/bitcoin-cli -signet -getinfo

.../...

Min tx relay fee rate (BTC/kvB): 0.00001000

Balances

  1.10997653 default settings
  0.00000000 disable private keys and blank
  0.00000000 disable private keys
  0.00000000 encrypted blank descriptor
  0.00000000 blank
498.77584341 ""
Warnings: This is a pre-release test build - use at your own risk - do not use for mining or merchant applications
```

- Could you please update the PR and commit titles to include "-getinfo"? i.e. something like `print -getinfo warnings only if warnings are present`

**Comment by theStack** [comment]
Concept ACK

**Comment by 1440000bytes** [comment]
> * It looks like there is now (a) an extra line between "Balances" and the balances, and (b) a line missing before the warnings, with both one balance and multiple balances. Example:

Sorry, missed this. Will do it this weekend.

> Could you please update the PR and commit titles to include "-getinfo"? i.e. something like print -getinfo warnings only if warnings are present

Done

**Comment by laanwj** [comment]
I don't know about leaving it out entirely. It might be reassuring to the user that there are, explicitly, no warnings. Or less reassuring (but impotent) that there could be warnings, but are none. Could format it different like `Warnings: (none)` or something like that if the empty field bothers you.

**Comment by jonatack** [comment]
> I don't know about leaving it out entirely. It might be reassuring to the user that there are, explicitly, no warnings. Or less reassuring (but important) that there could be warnings, but are none. Could format it different like `Warnings: (none)` or something like that if the empty field bothers you.

Good point (and probably a simpler change to implement).

**Comment by 1440000bytes** [comment]
> I don't know about leaving it out entirely. It might be reassuring to the user that there are, explicitly, no warnings. Or less reassuring (but important) that there could be warnings, but are none. Could format it different like `Warnings: (none)` or something like that if the empty field bothers you.

Done in last commit. Update pull request title and description.

File: src/bitcoin-cli.cpp [hunk 2]
```diff
@@ -1059,7 +1059,13 @@ static void ParseGetInfoResult(UniValue& result)
         result_string += "\n";
     }
 
-    result_string += strprintf("%sWarnings:%s %s", YELLOW, RESET, result["warnings"].getValStr());
+    const std::string warnings{result["warnings"].getValStr()};
+    if (warnings.empty()) {
+        result_string += strprintf("%sWarnings: %s(none)", YELLOW, RESET);
```
Comment (jonatack) [inline src/bitcoin-cli.cpp:1064]:
```suggestion
        result_string += strprintf("%sWarnings:%s (none)", YELLOW, RESET);
```
------
alternatively, could simplify

```diff
-    if (warnings.empty()) {
-        result_string += strprintf("%sWarnings: %s(none)", YELLOW, RESET);
-    } else {
-        result_string += strprintf("%sWarnings:%s %s", YELLOW, RESET, warnings);
-    }
+    result_string += strprintf("%sWarnings:%s %s", YELLOW, RESET, warnings.empty() ? "(none)" : warnings);
```

Comment (1440000bytes) [reply src/bitcoin-cli.cpp]: Done in last commit.

**Comment by jonatack** [comment]
ACK 0cea7b10f1180e9993c14473e1a3b6525ef6ba01

image

**Comment by laanwj** [comment]
Tested ACK 0cea7b10f1180e9993c14473e1a3b6525ef6ba01

**Pull request merged**